File: app/src/main/java/com/knd/duantotnghiep/duantotnghiep/core/BaseFragment.java
package com.knd.duantotnghiep.duantotnghiep.core;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewbinding.ViewBinding;

public abstract class BaseFragment<VB extends ViewBinding> extends Fragment {
    protected VB binding;
    private int requestCode = 0;

    @LayoutRes
    private final int layout;

    public BaseFragment(@LayoutRes int layout) {
        this.layout = layout;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(layout, container, false);
        binding = getViewBinding(rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        Log.d("palsplad", "onViewCreated");
        super.onViewCreated(view, savedInstanceState);
        initData();
        initView();
        initObserver();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d("palsplad", "onDestroy");
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.d("palsplad", "onResume");
    }

    public void showMessage(String string) {
        Toast.makeText(requireContext(), string, Toast.LENGTH_SHORT).show();
    }

    public abstract void initObserver();

    public abstract VB getViewBinding(View view);

    public void initView() {
    }

    public void initData() {
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/knd/duantotnghiep/duantotnghiep/core/BaseFragment.kt
package com.knd.duantotnghiep.duantotnghiep.core

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Message
import android.util.Log
 import android.view.View

import androidx.activity.result.ActivityResult

import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import androidx.viewbinding.ViewBinding
import com.google.android.gms.auth.api.identity.Identity

import com.google.android.gms.common.api.ApiException

abstract class BaseFragment<VB : ViewBinding>(layout: Int) : Fragment(layout) {
    private var _binding: VB? = null
    private var requestCode = 0
    open val binding get() = _binding!!


    private fun resultSignInGoogle(result: ActivityResult) {
        try {
            val credential = Identity.getSignInClient(requireActivity())
                .getSignInCredentialFromIntent(result.data)
            val idToken = credential.googleIdToken

            if (idToken != null) {
                fragmentResultCompletion(idToken, true)

            } else {
                fragmentResultCompletion("Error", false)
            }
        } catch (e: ApiException) {
            fragmentResultCompletion(e.localizedMessage, false)
        }
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = getViewBinding(view)
        initData()
        initView()
        initObserver()
    }


    open var fragmentResultCompletion: (String?, Boolean) -> Unit = { param, s -> }


    fun launchFragment(id: Int) = try {
        findNavController().navigate(id)
    } catch (e: Exception) {
        Log.d("okokok", e.localizedMessage)
    }

    fun popBackStack() = findNavController().popBackStack()

    open fun initObserver() {}

    abstract fun getViewBinding(view: View): VB
    open fun initView() {}
    open fun initData() {}
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}